### PR TITLE
Add Arc and Pie Drawing and Plotting Features

### PR DESCRIPTION
**Pull request by varphone**
## Summary

This PR introduces several new features related to drawing and plotting arcs and pies in the `egui` ecosystem.

## Changes

- `epaint`: Introduced `ArcPieShape`, a new shape that represents an arc or a pie slice with specified radius, angle, fill, and stroke properties.
- `egui`: Added `Painter::{arc, pie}` methods, which simplify the process of drawing arcs and pies.
- `egui_plot`: Introduced three new plot types: `ArcLine`, `Pie`, and `PieChart`. `ArcLine` renders a simple arc line, `Pie` renders a single pie slice, and `PieChart` renders a pie chart with data and labels.
- `egui_demo_lib`: Added demonstrations for `ArcPieShape`, `ArcLine`, `Pie`, and `PieChart`. These demos showcase how to use these new features in a variety of contexts.

## Impact

These changes enhance the capabilities of the `egui` ecosystem by providing users with more options for drawing and plotting.
The new demos also serve as valuable examples for users to learn from.

[screencast-egui_demo_app-2024-04-27.webm](https://github.com/emilk/egui/assets/1274171/f91e9600-aff5-4bba-9c79-f5d2596d1b0e)

### CODE REVIEW

File: crates/egui/src/painter.rs [hunk 1]
```diff
@@ -297,6 +297,38 @@ impl Painter {
...
     }
 
+    /// Paints an arc line.
+    pub fn arc(
+        &self,
+        center: Pos2,
+        radius: f32,
+        start_angle: f32,
+        end_angle: f32,
+        stroke: impl Into<Stroke>,
```
Comment (murl-digital) [inline crates/egui/src/painter.rs:307]: Please change this to `Into<PathStroke>`, this was recently added in epaint

File: crates/epaint/src/shape.rs [hunk 2]
```diff
@@ -233,6 +236,80 @@ impl Shape {
...
+    /// ```no_run
+    /// # use epaint::{pos2, Color32, Shape, Stroke};
+    /// let arc = Shape::arc(pos2(100.0, 100.0), 50.0, 0.0, std::f32::consts::PI, Stroke::new(3.0, Color32::RED));
+    /// ```
+    pub fn arc(
+        center: Pos2,
+        radius: f32,
+        start_angle: f32,
+        end_angle: f32,
+        stroke: impl Into<Stroke>,
```
Comment (murl-digital) [inline crates/epaint/src/shape.rs:264]: Same here, change this to `Into<PathStroke>`, this was recently added in epaint

File: crates/epaint/src/shape.rs [hunk 3]
```diff
@@ -233,6 +236,80 @@ impl Shape {
...
+    /// # use epaint::{pos2, Color32, Shape, Stroke};
+    /// let pie = Shape::pie(pos2(100.0, 100.0), 50.0, 0.0, std::f32::consts::PI, Color32::BLUE, Stroke::new(3.0, Color32::RED));
+    /// ```
+    pub fn pie(
+        center: Pos2,
+        radius: f32,
+        start_angle: f32,
+        end_angle: f32,
+        fill: impl Into<Color32>,
+        stroke: impl Into<Stroke>,
```
Comment (murl-digital) [inline crates/epaint/src/shape.rs:301]: `Into<PathStroke>`

**Review by murl-digital — CHANGES_REQUESTED** [review]
Only thing I saw was using Stroke instead of PathStroke for lines.

File: crates/epaint/src/tessellator.rs [hunk 4]
```diff
@@ -1440,6 +1440,14 @@ impl Tessellator {
 
         // If the arc is a full circle, we can just use the circle function.
         if (end_angle - start_angle).abs() >= std::f32::consts::TAU {
+            let stroke_color = match stroke.color {
+                ColorMode::Solid(color) => color,
+                ColorMode::UV(callback) => {
+                    // We can't use the callback here, so we just use the center of the circle.
```
Comment (murl-digital) [inline crates/epaint/src/tessellator.rs:1446]: Maybe this behavior should be documented somewhere? I'm also not fully sure if you can't use the callback here, could you elaborate?

**Review by murl-digital — COMMENTED** [review]
Thanks for pushing that change! I'm a bit tired right now so I'll look over everything else when I get the chance. (feel free to poke me)

File: crates/epaint/src/tessellator.rs (hunk 4 above)
Comment (varphone) [reply crates/epaint/src/tessellator.rs:1446]: The entire pie is rendered as a circle, but `CircleShape` is not compatible with `PathStroke`. 
As a workaround, I use a callback to obtain the center color from `PathStroke`. 
This issue should be addressed once `CircleShape` supports `PathStroke`.

Comment (murl-digital) [reply crates/epaint/src/tessellator.rs:1446]: Ah, I understand. Thanks for elaborating!

Im hoping that the underlying `ColorMode` will be used for more shapes in the future, but for the time being it is only for paths. If you want, I can look into opening a PR to support `PathStroke` for more shapes

**Comment by enomado** [comment]
Take a look to a bezier curve arc approximation methods, it could be faster(may be)

https://docs.rs/bezier-nd/latest/bezier_nd/struct.Bezier.html#method.arc
https://michael-f-bryan.github.io/arcs/kurbo/struct.Arc.html

Example: https://github.com/enomado/smitters/blob/main/src/egui.rs#L6

**Comment by murl-digital** [comment]
> Take a look to a bezier curve arc approximation methods, it could be faster(may be)

It could also be more accurate, if you'd like an example, I have a duplicate PR: https://github.com/emilk/egui/pull/4426

**Comment by emilk** [comment]
`egui_plot` has recently been moved to its own repository, at https://github.com/emilk/egui_plot

This will hopefully speed up its development by having more reviewers and maintainers.

Please re-open this PR at https://github.com/emilk/egui_plot/pulls

See also:
* https://github.com/emilk/egui/issues/4705
* https://github.com/emilk/egui/pull/4828

**Comment by murl-digital** [comment]
@emilk, I'd categorize this as an epaint feature request, or at least a part of it

**Comment by Skgland** [comment]
Currently looking into how to make a pie chart with egui and stumbled upon this.

Having circle arcs available in epaint/egui would make this easier as I wouldn't have to approximate a circle with a bezier curve. Obviously having a ready made pie chart in egui_plot would be even better.

**Pull request closed without merging**